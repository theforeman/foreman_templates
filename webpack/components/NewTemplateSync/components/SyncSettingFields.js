import React from 'react';
import { upperFirst } from 'lodash';
import PropTypes from 'prop-types';

import SyncSettingField from './SyncSettingField';

const SyncSettingsFields = ({
  importSettings,
  exportSettings,
  syncType,
  resetField,
  formProps: { isSubmitting },
}) => {
  const addRequiredToSetting = setting =>
    setting.name === 'repo'
      ? setting.merge({
          required: true,
        })
      : setting;

  const modifyDescription = (setting, type) => {
    if (setting.description) {
      let split = setting.description.split('. ');
      if (setting.name === 'repo' && type !== 'export') {
        split = split.slice(0, split.length - 1);
      }

      split = split.join('.<br>');
      return setting.set('description', split);
    }
    return setting;
  };

  const specializeDescription = (setting, type) =>
    setting.set(
      'description',
      upperFirst(setting.description.replace(/import\/export/i, type))
    );

  const settingsAry = syncType === 'import' ? importSettings : exportSettings;

  return (
    <React.Fragment>
      {settingsAry
        .map(addRequiredToSetting)
        .map(setting => modifyDescription(setting, syncType))
        .map(setting => specializeDescription(setting, syncType))
        .map(setting => (
          <SyncSettingField
            setting={setting}
            syncType={syncType}
            key={setting.name}
            disabled={isSubmitting}
            resetField={resetField}
          />
        ))}
    </React.Fragment>
  );
};

SyncSettingsFields.propTypes = {
  importSettings: PropTypes.array.isRequired,
  exportSettings: PropTypes.array.isRequired,
  syncType: PropTypes.string.isRequired,
  resetField: PropTypes.func.isRequired,
  formProps: PropTypes.object,
};

SyncSettingsFields.defaultProps = {
  formProps: {},
};

export default SyncSettingsFields;
