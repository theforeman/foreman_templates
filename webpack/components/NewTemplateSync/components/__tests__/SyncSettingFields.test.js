import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import SyncSettingFields from '../SyncSettingFields';

import {
  importSettings,
  exportSettings,
} from '../../__fixtures__/templateSyncSettings.fixtures';

const noop = () => {};

const commonFixtures = {
  importSettings,
  exportSettings,
  resetField: noop,
  formProps: { isSubmitting: false },
  disabled: false,
};

const fixtures = {
  'should show export settings': {
    syncType: 'export',
    validationData: {},
    ...commonFixtures,
  },
  'should show import settings': {
    syncType: 'import',
    validationData: { repo: ['http', 'https'] },
    ...commonFixtures,
  },
};

describe('SyncSettingFields', () =>
  testComponentSnapshotsWithFixtures(SyncSettingFields, fixtures));
