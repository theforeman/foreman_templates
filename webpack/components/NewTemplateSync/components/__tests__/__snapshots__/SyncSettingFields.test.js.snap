// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncSettingFields should show export settings 1`] = `
<Fragment>
  <SyncSettingField
    disabled={false}
    key="filter"
    resetField={[Function]}
    setting={
      Object {
        "description": "Import or export names matching this regex (case-insensitive; snippets are not filtered)",
        "id": 47,
        "name": "filter",
        "settingsType": "string",
        "value": "",
      }
    }
    syncType="export"
  />
  <SyncSettingField
    disabled={false}
    key="negate"
    resetField={[Function]}
    setting={
      Object {
        "description": "Negate the filter for export",
        "id": 48,
        "name": "negate",
        "settingsType": "bool",
        "value": false,
      }
    }
    syncType="export"
  />
  <SyncSettingField
    disabled={false}
    key="repo"
    resetField={[Function]}
    setting={
      Object {
        "description": "Target path to import and export.<br>Different protocols can be used, for example /tmp/dir, git://example.com, https://example.com, ssh://example.com.<br>When exporting to /tmp, note that production deployments may be configured to use private tmp.",
        "id": 49,
        "name": "repo",
        "required": true,
        "settingsType": "string",
        "value": "https://github.com/theforeman/community-templates.git",
      }
    }
    syncType="export"
  />
</Fragment>
`;

exports[`SyncSettingFields should show import settings 1`] = `
<Fragment>
  <SyncSettingField
    disabled={false}
    key="associate"
    resetField={[Function]}
    setting={
      Object {
        "description": "Associate templates to OS, organization and location",
        "id": 45,
        "name": "associate",
        "selection": Array [
          Object {
            "label": "New",
            "value": "new",
          },
          Object {
            "label": "Never",
            "value": "never",
          },
          Object {
            "label": "Always",
            "value": "always",
          },
        ],
        "settingsType": "string",
        "value": "new",
      }
    }
    syncType="import"
  />
  <SyncSettingField
    disabled={false}
    key="force"
    resetField={[Function]}
    setting={
      Object {
        "description": "Should importing overwrite locked templates?",
        "id": 46,
        "name": "force",
        "settingsType": "bool",
        "value": false,
      }
    }
    syncType="import"
  />
</Fragment>
`;
